sycl/include/sycl/detail/vector_arith.hpp

```diff
@@ -24,49 +24,166 @@ template <typename DataT, int NumElem> class __SYCL_EBO vec;
-    } else {                                                                   \
-      auto ExtVecLhs = sycl::bit_cast<typename vec_t::vector_t>(Lhs);          \
-      auto ExtVecRhs = sycl::bit_cast<typename vec_t::vector_t>(Rhs);          \
-      Ret = vec<DataT, NumElements>(ExtVecLhs BINOP ExtVecRhs);                \
-      if constexpr (std::is_same_v<DataT, bool> && CONVERT) {                  \
-        vec_arith_common<bool, NumElements>::ConvertToDataT(Ret);              \
-      }                                                                        \
-    }                                                                          \
-    return Ret;                                                                \
+  static constexpr bool is_host = false;
+#else
+  static constexpr bool is_host = true;
+#endif
+
+  template <typename BinOp, typename Self>
+  static constexpr auto apply_binary_op(const Self &Lhs, const Self &Rhs) {
+    static_assert(is_vec_v<Self>);
+    using element_type = typename Self::element_type;
+    constexpr int N = Self::size();
+    constexpr bool is_logical =
+        check_type_in_v<BinOp, std::equal_to<void>, std::not_equal_to<void>,
+                        std::less<void>, std::greater<void>,
+                        std::less_equal<void>, std::greater_equal<void>,
+                        std::logical_and<void>, std::logical_or<void>>;
+    using result_t = std::conditional_t<
+        is_logical, vec<fixed_width_signed<sizeof(element_type)>, N>, Self>;
+
+    BinOp Op{};
+    if constexpr (is_host ||
+                  std::is_same_v<element_type, ext::oneapi::bfloat16>) {
+      result_t res{};
+      for (size_t i = 0; i < N; ++i)
+        if constexpr (is_logical)
+          res[i] = Op(Lhs[i], Rhs[i]) ? -1 : 0;
+        else
+          res[i] = Op(Lhs[i], Rhs[i]);
+      return res;
+    } else {
+      using vector_t = typename Self::vector_t;
+      auto res = [&](auto x, auto y) {
+        // Workaround for https://godbolt.org/z/P6daoG6jW.
+        if constexpr (std::is_same_v<BinOp, std::equal_to<void>>)
+          return x == y;
+        else if constexpr (std::is_same_v<BinOp, std::not_equal_to<void>>)
+          return x != y;
+        else if constexpr (std::is_same_v<BinOp, std::less<void>>)
+          return x < y;
+        else if constexpr (std::is_same_v<BinOp, std::less_equal<void>>)
+          return x <= y;
+        else if constexpr (std::is_same_v<BinOp, std::greater<void>>)
+          return x > y;
+        else if constexpr (std::is_same_v<BinOp, std::greater_equal<void>>)
+          return x >= y;
+        else
+          return Op(x, y);
+      }(bit_cast<vector_t>(Lhs), bit_cast<vector_t>(Rhs));
+      if constexpr (is_logical) {
+        if constexpr (N == 1) {
+          // vector_t for one-element vector is just scalar with normal C++
+          // semantics, need to align it with vector behavior:
+          return result_t{(typename result_t::vector_t)res * -1};
```

Comment:
```suggestion
          return result_t{static_cast<typename result_t::vector_t>(res) * -1};
```
Let's use `static_cast` over C-style cast.

Reply:
IIRC from similar experiments in the past, `static_cast` won't help and we'll have to use C-style one. I'll try once again and add a comment about that if that's correct.